x-pack/plugins/security_solution/public/detections/containers/detection_engine/alerts/use_user_profiles.tsx

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { UserProfileWithAvatar } from '@kbn/user-profile-components';
+import { useEffect, useState } from 'react';
+import { useAppToasts } from '../../../../common/hooks/use_app_toasts';
+import { fetchUserProfiles } from './api';
+import { USER_PROFILES_FAILURE } from './translations';
+
+interface UserProfilesStatus {
+  loading: boolean;
+  userProfiles: UserProfileWithAvatar[];
+}
+
+export const useUserProfiles = (): UserProfilesStatus => {
+  const [loading, setLoading] = useState(false);
+  const [users, setUsers] = useState<UserProfileWithAvatar[]>([]);
+  const { addError } = useAppToasts();
+
+  useEffect(() => {
+    // isMounted tracks if a component is mounted before changing state
+    let isMounted = true;
+    setLoading(true);
+    const fetchData = async () => {
+      try {
+        const usersResponse = await fetchUserProfiles();
```

Comment:
Fetching all user profiles here then filtering them on the client side seems dangerous. Is it possible to only fetch the profiles for IDs on the current page of the alerts table?

I have a similar need for the `kibana.alert.workflow_user` field, I'd like to display the username rather than the profile ID.

Reply:
We need to present user the list of all available user to pick assignees.

<img width="699" alt="Screenshot 2023-10-04 at 18 17 57" src="https://github.com/elastic/kibana/assets/2700761/bc8ddf03-9585-4a0a-82be-088eae10d5b9">

This mimics similar functionality in `cases` plugin (`x-pack/plugins/cases/public/containers/user_profiles/use_suggest_user_profiles.ts`).

One this what we might need to improve is required privileges to fetch user profiles, right now I allow to fetch profiles if user can login (`x-pack/plugins/security_solution/server/lib/detection_engine/routes/signals/get_user_profiles_route.ts`):

```
const users = await security.userProfiles.suggest({
            name: '',
            dataPath: 'avatar',
            requiredPrivileges: {
              spaceId,
              privileges: {
                kibana: [security.authz.actions.login],
              },
            },
          });
```

Reply:
Do we need to pass the `name` in to get the correct suggestions?

I saw that this hook is also used in `getRenderCellValueHook` though to convert from profile UID to username for display in the table. In that case wouldn't we need to ensure that we're fetching the correct profiles by ID when rendering the table? By default it appears the suggest API only returns 10 users.

Reply:
Right, I was playing around with three users in my environment that's why I did not hit the limit of 10 suggestions. I will adjust suggest APIs and also will add one more route to be able to fetch specific user profiles by passing their ids.

Do you think that will cover all the cases for `kibana.alert.workflow_user` as well?

Reply:
Yeah I think the solution here should cover what we'll need for `workflow_user`. There's a `UserProfileAPIClient` in `public` already though, I wonder if we can use that instead of making our own new APIs? Something like 
```
import type { UserProfileWithAvatar } from '@kbn/user-profile-components';
import { useEffect, useState } from 'react';
import { useAppToasts } from '../../../../common/hooks/use_app_toasts';
import { useKibana } from '../../../../common/lib/kibana';
import { USER_PROFILES_FAILURE } from './translations';

interface UserProfilesStatus {
  loading: boolean;
  userProfiles: UserProfileWithAvatar[];
}

// Gather all uids from one page of the alerts table and fetch those userprofiles in one request
export const useUserProfiles = ({ uids }: { uids: string[] }): UserProfilesStatus => {
  const [loading, setLoading] = useState(false);
  const [users, setUsers] = useState<UserProfileWithAvatar[]>([]);
  const { addError } = useAppToasts();
  const userProfiles = useKibana().services.security.userProfiles;

  useEffect(() => {
    // isMounted tracks if a component is mounted before changing state
    let isMounted = true;
    setLoading(true);
    const fetchData = async () => {
      try {
        const usersResponse = await userProfiles.bulkGet({ uids: new Set(uids) });
        if (isMounted) {
          setUsers(usersResponse);
        }
      } catch (error) {
        addError(error.message, { title: USER_PROFILES_FAILURE });
      }
      if (isMounted) {
        setLoading(false);
      }
    };
    fetchData();
    return () => {
      // updates to show component is unmounted
      isMounted = false;
    };
  }, [addError, uids, userProfiles]);
  return { loading, userProfiles: users };
};
```
for `bulkGet`, there's also a `suggest` method on the client so we might be able to use this.

Reply:
ohh, seems like I overcomplicated things around `UserProfileAPIClient` APIs, I use those APIs, but on the server side of security solution :-) I will have a look and double check that it works correctly without an unnecessary trip to our server first.

Reply:
Seems like we need to implement our own API at least for `suggest` call. Here is the note from the `UserProfileAPIClient`:
```
Note: This endpoint is not provided out-of-the-box by the platform. You need to expose your own
version within your app. An example of how to do this can be found in:
`examples/user_profile_examples/server/plugin.ts`
```